### DISCUSSION TITLE
styling pandas dataframes?

### DISCUSSION THREAD

**Discussion by fdion** [executablebooks/meta#308] [Q&A]
Wondering if there's a way to style pandas dataframes when using the pdflatex builder (with the jupyterbook_latex sphinx extension)?

Pandas does have a `.to_latex()` but if I use `display(Latex(df.to_latex()))`, the process doesn't complete.

I did add `booktabs` to my `_config.yml` file:
```
latex:
  latex_engine              : xelatex
  use_jupyterbook_latex     : true
  latex_elements:
      preamble   : |
       \usepackage{booktabs}
```

**EDIT**: It should have added `\usepackage{booktabs}` to the .tex file, but didn't. I manually added it and ran `make` and the PDF generated like it should.

So, apparently I don't have quite the right syntax for the `_config.yml` file, if anybody has a suggestion as to the proper way to set it up?

**Comment by fdion** [comment]
Looking at the code here:

https://github.com/executablebooks/jupyter-book/blob/8be06d8b8920d4967737d7ffec91af691d7100b1/jupyter_book/config.py#L312

it is obvious this is not handled by the `latex` section. Also, looking at:
https://github.com/executablebooks/jupyter-book/blob/8be06d8b8920d4967737d7ffec91af691d7100b1/docs/_config.yml#L76

jupyterbook documentation itself uses a latex_elements section under sphinx, as you would in a sphinx `conf.py`. Ok, let's try that:

It doesn't add booktabs to the tex file...

**Reply by fdion** [reply]
I was able to run (finally) `jb config sphinx bookdir/`. I had to disable `only_build_toc_files: true` to get this to run.

Then the issue became obvious. Needed to move `latex_elements` under `config`, under the `sphinx` heading:

```
sphinx:
    extra_extensions:
        - jupyterbook_latex
    config:
      latex_elements:
        preamble   : |
          \usepackage{booktabs}
```

Figured someone else might run into this (especially the is a directory error if only_build_toc_files is true...

**Reply by chrisjsewell** [reply]
Always good when someone answers their own question thanks @fdion 😄 

> especially the[re] is a directory error if `only_build_toc_files` is true

Could you print the error here?
Note this may be fixed by https://github.com/executablebooks/jupyter-book/pull/1293, since `only_build_toc_files` will now only exclude document files (https://github.com/executablebooks/sphinx-external-toc/pull/9)

**Reply by fdion** [reply]
Here's what I get if I try to run `jb config sphinx bookdir/` when using `only_build_toc_files: true`:
```
...
  File "/home/xxx/site-packages/jupyter_book/commands/__init__.py", line 488, in sphinx
    sphinx_config, config_meta = get_final_config(
  File "/home/xxx/site-packages/jupyter_book/config.py", line 131, in get_final_config
    newly_excluded = _get_files_outside_toc(toc, sourcedir, excluded_patterns)
  File "/home/xxx/site-packages/jupyter_book/config.py", line 401, in _get_files_outside_toc
    toc_yaml = yaml.safe_load(toc.read_text(encoding="utf8"))
  File "/home/xxx/pathlib.py", line 1255, in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
  File "/home/xxx/pathlib.py", line 1241, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
IsADirectoryError: [Errno 21] Is a directory: '.'
```

This is on `jupyter-book 0.10.1`.  I saw `0.10.2`, just updated and...

It is now fixed.

**Reply by chrisjsewell** [reply]
thanks!

**Reply by a7p** [reply]
I'm having the same question, but am not getting the answers here.

I'm glueing a pandas DataFrame and it get's displayed in my pdf-output, but the rendering is _very_ rudimentary and has no resemblance to what I'd get with `df.to_latex()`. Is it even possible to get styled DataFrames in pdf-output? I'm kind of unsure what to expect here.
